src/runtime/virtcontainers/qemu_amd64.go

```diff
@@ -280,11 +319,297 @@ func (q *qemuAmd64) appendProtectionDevice(devices []govmmQemu.Device, firmware,
 				File:            firmware,
 				CBitPos:         cpuid.AMDMemEncrypt.CBitPosition,
 				ReducedPhysBits: cpuid.AMDMemEncrypt.PhysAddrReduction,
+				SevPolicy:       policy,
+				CertFilePath:    godhPath,
+				SessionFilePath: sessionFilePath,
+				KernelHashes:    true,
 			}), "", nil
-	case noneProtection:
-		return devices, firmware, nil
+	} else {
+		return append(devices,
+			govmmQemu.Object{
+				Type:            govmmQemu.SEVGuest,
+				ID:              "sev",
+				Debug:           false,
+				File:            firmware,
+				CBitPos:         cpuid.AMDMemEncrypt.CBitPosition,
+				ReducedPhysBits: cpuid.AMDMemEncrypt.PhysAddrReduction,
+				SevPolicy:       policy,
+			}), "", nil
+	}
+}
 
-	default:
-		return devices, "", fmt.Errorf("Unsupported guest protection technology: %v", q.protection)
+// setup prelaunch attestation
+func (q *qemuArchBase) setupSEVGuestAttestation(ctx context.Context,
+	proxy string,
+	policy uint32) (string, error) {
+
+	logger := virtLog.WithField("subsystem", "SEV attestation")
+	logger.Info("Set up prelaunch attestation")
+
+	certChainBin, err := os.ReadFile(filepath.Join(sevAttestationWorkingDir, sevAttestationCertChainName))
+	if err != nil {
+		log.Fatalf("cert chain not found: %v", err)
+	}
+
+	certChain := b64.StdEncoding.EncodeToString([]byte(certChainBin))
+
+	conn, err := grpc.Dial(proxy, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		log.Fatalf("did not connect: %v", err)
+	}
+
+	client := pb.NewKeyBrokerServiceClient(conn)
+	ctx, cancel := context.WithTimeout(context.Background(), sevAttestationGrpcTimeout)
+	defer cancel()
+
+	request := pb.BundleRequest{
+		CertificateChain: string(certChain),
+		Policy:           policy,
+	}
+	bundleResponse, err := client.GetBundle(ctx, &request)
+	if err != nil {
+		log.Fatalf("did not connect: %v", err)
+	}
+
+	attestationId := bundleResponse.LaunchId
+	attestationDataPath := filepath.Join(sevAttestationWorkingDir, attestationId)
+	err = os.MkdirAll(attestationDataPath, os.ModePerm)
+	if err != nil {
+		log.Fatalf("sev attestation directory not created: %v", err)
+	}
+
+	godhPath := filepath.Join(attestationDataPath, sevAttestationGodhName)
+	sessionFilePath := filepath.Join(attestationDataPath, sevAttestationSessionFileName)
+
+	err = os.WriteFile(godhPath, []byte(bundleResponse.GuestOwnerPublicKey), 0777)
+	if err != nil {
+		log.Fatal(err)
+	}
+	err = os.WriteFile(sessionFilePath, []byte(bundleResponse.LaunchBlob), 0777)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	return attestationId, nil
+}
+
+type guidLE [16]byte
+
+// The following definitions must be identical to those in QEMU target/i386/sev.c
```

Comment:
This sounds a bit fragile. _Could_ they ever change?

Reply:
If we ever want to change this in QEMU, we will do this is in a backward compatible way by adding a version option; something like:

    -object sev-guest,....,kernel-hashes=on,kernel-hashes-version=2

So as long as kata passes a command-line without `kernel-hashes-version=2` then QEMU will generate the same memory structure as appears here. Later when we want to add the (hypothetical) kernel-hashes v2 format, we can add this to kata as well (maybe falling back to v1 if QEMU is too old, etc.).

We have similar issue in the interface between QEMU and OVMF (QEMU generates this table and OVMF reads it and verifies the hashes inside the encrypted VM; so any change in format needs to be coordinated and backward-compatible in both ways: old QEMU + new OVMF and vice versa)).

Reply:
These definitions have been moved out of `qemu_amd64.go` and into an SEV utility package: `pkg/sev/sev.go`